colossalai/gemini/memory_tracer/memstats_collector.py

```diff
@@ -150,3 +158,61 @@ def sample_model_data(self) -> None:
     @property
     def cuda_margin_mem(self) -> float:
         return colo_device_memory_capacity(get_current_device()) - max(self.overall_mem_stats('cuda'))
+
+
+class MemStatsCollectorStatic(MemStatsCollector):
+    """
+    A Static Memory statistic collector.
+    """
+
+    def __init__(self, module: nn.Module) -> None:
+        super().__init__()
+        self.module = module
+
+    def init_mem_stats(self, *inputs):
+        self.module.cpu()
+        self.module.train()
+        # data = MetaTensor(torch.rand(int(TM_BATCH_SIZE), 3, 224, 224, device='meta'), fake_device='cuda:0')
+        data = [MetaTensor(torch.rand(inp.shape, device='meta'), fake_device='cpu') for inp in inputs]
+        gm = symbolic_trace(self.module)
+        interp = MetaInfoProp(gm)
+        interp.propagate(data)
+        for node in gm.graph.nodes:
+            self._non_model_data_cuda_list.append(node.meta["fwd_mem_tmp"] + node.meta['fwd_mem_out'])
+        for node in gm.graph.nodes.__reversed__():
+            self._non_model_data_cuda_list.append(node.meta["bwd_mem_tmp"] + node.meta['bwd_mem_out'])
```

Comment:
For this part of backward, we might have to consider the freed forward tensors. I will dive deep here to check the whole system of gemini.